### PR TITLE
Convert `TrapCode` to a single byte

### LINKED ISSUE #9310: Replace `cranelift_codegen::ir::TrapCode` with a `u8`
We currently have an `enum` of various trap codes, but they are all pretty Wasm-/Wasmtime-specific, rather than things that general Cranelift users will want. And they certainly do not cover everything that general Cranelift users might want.

We should replace `cranelift_codegen::ir::TrapCode` with a `u8` that is simply plumbed through cranelift and whose meaning is determined fully by the CLIF producer. Or perhaps instead of simply a `u8`, a `pub struct TrapCode(pub u8)`, just so that it is that much harder to pass the wrong value around when we are shuffling these about inside the compiler.

This will also let us stop duplicating trap codes between `cranelift_codegen::ir::TrapCode` and `wasmtime_environ::Trap`, which will be a nice clean up.

Note: fixing this issue will mostly involve chasing down compile errors throughout Cranelift and Wasmtime.

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit refactors the representation of
`cranelift_codegen::ir::TrapCode` to be a single byte. The previous enumeration is replaced with an opaque byte-sized structure. Previous variants that Cranelift uses internally are now associated `const` values on `TrapCode` itself. For example `TrapCode::IntegerOverflow` is now `TrapCode::INTEGER_OVERFLOW`. All non-Cranelift traps are now removed and exclusively live in the `wasmtime-cranelift` crate now.

The representation of a `TrapCode` is now:

* 0 - invalid, used in `MemFlags` for "no trap code"
* 1..256-N - user traps
* 256-N..256 - built-in Cranelift traps (it uses N of these)

This enables embedders to have 255-N trap codes which is more than enough for Wasmtime for example. Cranelift reserves a few built-in codes for itself which shouldn't eat too much into the trap space. Additionally if Cranelift needs to grow a new trap it can do so pretty easily too.

The overall intent of this commit is to reduce the coupling of Wasmtime and Cranelift further and generally refactor Wasmtime to use user traps more often. This additionally shrinks the size of `TrapCode` for storage in various locations, notably it can now infallibly be represented inside of a `MemFlags`.

Closes #9310

### CODE REVIEW

File: crates/cranelift/src/lib.rs [hunk 1]
```diff
@@ -33,8 +33,24 @@ mod func_environ;
 mod gc;
 mod translate;
 
-/// User trap code for debug assertions we emit in our JIT code.
-const DEBUG_ASSERT_TRAP_CODE: u16 = u16::MAX;
+const TRAP_DEBUG_ASSERT: TrapCode = TrapCode::user(1);
```
Comment (fitzgen) [inline crates/cranelift/src/lib.rs:36]: What do you think about renaming this to `TRAP_INTERNAL_ASSERT` while we're here? We emit this not just in debug builds but also in release builds for certain asserts, so the "debug" nomenclature has been kind of nagging at me recently.

**Review by fitzgen — APPROVED** [review]
Nice! A few nitpicks inline below.

File: cranelift/codegen/src/ir/memflags.rs [hunk 2]
```diff
@@ -343,58 +342,19 @@ impl MemFlags {
...
-            Some(TrapCode::BadConversionToInteger) => 0b1000,
-            Some(TrapCode::UnreachableCodeReached) => 0b1001,
-            Some(TrapCode::Interrupt) => 0b1010,
-            Some(TrapCode::NullReference) => 0b1011,
-            Some(TrapCode::ArrayOutOfBounds) => 0b1100,
-            Some(TrapCode::AllocationTooLarge) => 0b1101,
-            None => 0b1111,
-
-            Some(TrapCode::User(_)) => panic!("cannot set user trap code in mem flags"),
+            Some(code) => code.as_raw() as u16,
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply cranelift/codegen/src/ir/memflags.rs:356]: Alas `const` rust thwarts this as `u16::from` isn't available in a `const` context

[the remark this answers is not in the document]
Comment (fitzgen) [reply cranelift/codegen/src/ir/memflags.rs:356]: Ah, gotcha

File: cranelift/codegen/src/ir/memflags.rs [hunk 3]
```diff
@@ -408,7 +368,7 @@ impl fmt::Display for MemFlags {
             None => write!(f, " notrap")?,
             // This is the default trap code, so don't print anything extra
             // for this.
-            Some(TrapCode::HeapOutOfBounds) => {}
+            Some(TrapCode::HEAP_OUT_OF_BOUNDS) => {}
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply cranelift/codegen/src/ir/memflags.rs:371]: Lack of trap though prints as `notrap`, so I think we're good?

**Pull request merged**